# Analisi sul data set Amazon E-commerce Products & Reviews

Il [dataset Amazon E-commers Products & Reviews](https://www.kaggle.com/datasets/lazylad99/amazon-e-commerce-product-and-review-dataset?select=reviews.csv) su Kaggle, contiene metadati dettagliati dei prodotti Amazon, recensioni dei clienti e punteggi di sentimenti. Questo insieme di dati è particolarmente indicato per analizzare il mercato e i comportamenti d'acquisto.

In [56]:
# import libraries

# data analisys
import pandas as pd
import numpy as np

# plotting/data visualization
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.dates import DateFormatter, AutoDateFormatter

## Comprensione dei Dati (EDA)

### Loading dataset

In [57]:
# load dataset
dfamazon = pd.read_csv('products_with_reviews_clean.csv')

# show first 10 rows of the dataset
display(dfamazon.head(10))

# show last 10 rows of the dataset
display(dfamazon.tail(10))

s_no                                         about_item        asin  \
0   146  SLIM FIT. Designed with a functional fit in mi...  B0006U6C0Q   
1    54  DESIGNS: Right sleeve with a pocket of the pol...  B09TK9248H   
2   103  Super-light HeatGear fabric delivers superior ...  B0874WD2B2   
3    56  100% Polyester, Quick Dry & Lightweight，Camo p...  B0C1SC8PR9   
4    43  High Performance Material: Polo shirts for men...  B0D38YFY3K   
5   103  Super-light HeatGear fabric delivers superior ...  B0874WD2B2   
6   153  Slim fit,simple style,basic jeans you need eve...  B01N58LDY4   
7    76  America's #1 Men's Underwear Brand - Source: C...  B086L1PM8V   
8    87  Named Best Anti-sweat undershirt by Men’s Heal...  B074XY7MTG   
9     9  Performance Fabric Blend: material blend of 95...  B0CXDL2NP6   

  availability                                  best_sellers_rank  \
0     In Stock  #1,493 in Clothing, Shoes & Jewelry (See Top 1...   
1     In Stock  #49,564 in Clothing, Shoes & Jewelry (See Top ...   
2     In Stock  #662 in Clothing, Shoes & Jewelry (See Top 100...   
3     In Stock  #44,945 in Clothing, Shoes & Jewelry (See Top ...   
4     In Stock  #32,242 in Clothing, Shoes & Jewelry (See Top ...   
5     In Stock  #662 in Clothing, Shoes & Jewelry (See Top 100...   
6     In Stock  #13,904 in Clothing, Shoes & Jewelry (See Top ...   
7     In Stock  #580 in Clothing, Shoes & Jewelry (See Top 100...   
8     In Stock  #1,936 in Clothing, Shoes & Jewelry (See Top 1...   
9     In Stock  #5,623 in Clothing, Shoes & Jewelry (See Top 1...   

             brand_name                                        breadcrumbs  \
0        Wrangler Store  Clothing, Shoes & Jewelry › Men › Clothing › J...   
1            ZITY Store  Clothing, Shoes & Jewelry › Men › Clothing › A...   
2    Under Armour Store  Clothing, Shoes & Jewelry › Men › Clothing › A...   
3           MoFiz Store  Clothing, Shoes & Jewelry › Men › Clothing › S...   
4         ZENGVEE Store  Clothing, Shoes & Jewelry › Men › Clothing › A...   
5    Under Armour Store  Clothing, Shoes & Jewelry › Men › Clothing › A...   
6  FREDD MARSHALL Store  Clothing, Shoes & Jewelry › Men › Clothing › J...   
7           Hanes Store  Clothing, Shoes & Jewelry › Men › Clothing › S...   
8           Hanes Store  Clothing, Shoes & Jewelry › Men › Clothing › U...   
9         OYGSieg Store  Clothing, Shoes & Jewelry › Men › Clothing › S...   

                             customer_review_summary default_variant0  \
0  Customers appreciate the jeans for their quali...   size:34W x 29L   
1  Customers appreciate the nice colors and comfo...     size:X-Large   
2  Customers find that the compression shirts fit...      size:Medium   
3  Customers appreciate the comfortable fit, nice...       size:Large   
4  Customers appreciate the shirts' fit, look, an...       size:Large   
5  Customers find that the compression shirts fit...      size:Medium   
6  Customers find the pants comfortable and stret...   size:32W x 32L   
7  Customers appreciate the shirts for their soft...      size:Medium   
8  Customers find these shirts soft, comfortable,...       size:Large   
9  Customers appreciate the shirts' fit, comfort,...     size:X-Large   

                                    default_variant1  ...       review_id  \
0                             color:Prewashed Indigo  ...  R103VSGVPRSZ6F   
1                                        color:Gnblg  ...  R103YA89AQKKT4   
2                                                NaN  ...  R105XJ6QWC0GUL   
3                            color:B-army Green Camo  ...  R1084B47TWDC0Z   
4                 color:1black Gray Green Navy White  ...  R10A6DXDSXNG8V   
5                                                NaN  ...  R10ATSAVN4XYOC   
6                                        color:Black  ...  R10GNORP27X12F   
7                            color:Assorted - 6 Pack  ...  R10KUUWU5UNOWP   
8                                        color:White  ...  R10QBRAFSV6HU

s_no                                         about_item        asin  \
6345   910  Versatile Design: Three sleep gowns, featuring...  B0744PN2MV   
6346   903  100% Organic Cotton: All Burt's Bees Baby Paja...  B08Q3H4W7T   
6347   912  ComfortableTwo piece outfit set includes hoodi...  B08FDXYMR5   
6348   938  Includes 5 onesies, Weave Type: KnitCotton rib...  B07QPQLW34   
6349   904  Boys Lightweight Jogger Pants:Waistband has a ...  B0C8NGNT8L   
6350   991  STYLE - Classic high rise taper jean sits at n...  B00021NY28   
6351   952  【STRETCHY FABRIC】 Men's ribbed polo shirt made...  B0DMSVLHJ1   
6352   988  SKINNY FIT: Experience a snug fit with these s...  B07J31WFNB   
6353   943  【Material】 :This men's button down short sleev...  B0DYSWTXH4   
6354   987  Material: 100% cottonLength: 4 MtrsIdeal for r...  B07L124QBK   

                            availability best_sellers_rank  \
6345                            In Stock               NaN   
6346                            In Stock               NaN   
6347  Only 1 left in stock - order soon.               NaN   
6348                            In Stock               NaN   
6349                            In Stock               NaN   
6350                            In Stock               NaN   
6351                            In Stock               NaN   
6352                            In Stock               NaN   
6353                            In stock               NaN   
6354  Only 5 left in stock - order soon.               NaN   

                         brand_name  \
6345  Simple Joys by Carter's Store   
6346         Burt's Bees Baby Store   
6347                    Rebey Store   
6348                  Brand: Gerber   
6349                 Brand: WaltZon   
6350        Gloria Vanderbilt Store   
6351                 COOFANDY Store   
6352        Amazon Essentials Store   
6353                  PORCLAY Store   
6354                 Brand: Generic   

                                            breadcrumbs  \
6345  Clothing, Shoes & Jewelry › Baby › Baby Girls ...   
6346  Clothing, Shoes & Jewelry › Baby › Baby Boys ›...   
6347  Clothing, Shoes & Jewelry › Baby › Baby Boys ›...   
6348  Clothing, Shoes & Jewelry › Baby › Baby Boys ›...   
6349  Clothing, Shoes & Jewelry › Boys › Clothing › ...   
6350  Clothing, Shoes & Jewelry › Women › Clothing ›...   
6351  Clothing, Shoes & Jewelry › Men › Clothing › S...   
6352  Clothing, Shoes & Jewelry › Women › Clothing ›...   
6353  Clothing, Shoes & Jewelry › Men › Clothing › S...   
6354  Clothing, Shoes & Jewelry › Men › Clothing › P...   

                                customer_review_summary  default_variant0  \
6345  Customers appreciate the softness, cozy design...   size:0-3 Months   
6346  Customers appreciate the softness, comfort, an...           size:3T   
6347  Customers find the coordinated outfit has a cu...  size:9-12 Months   
6348  Customers find the leotards have a cute design...   size:0-3 Months   
6349  Customers appreciate the pants for their soft,...        size:13-14   
6350  Customers appreciate the pants' style, comfort...     size:14 Short   
6351                                                NaN        size:Small   
6352  Customers appreciate the pants for their comfo...           size:10   
6353                                                NaN        size:Small   
6354                                                NaN               NaN   

            default_variant1  ...      review_id helpful_vote_count  \
6345        color:Grey/White  ...  RXRJ80VKII5OW                5.0   
6346  color:Happy Herbivores  ...  RXZ8I5PXSG41U                0.0   
6347              color:Grey  ...  RY3FN036ESPPR                0.0   
6348        color:Hello Bear  ...  RY7256WOZWMNK                0.0   
6349       color:Wz3801black  ...  RYCDI8EKDNLI3                0.0   
6350                     NaN  ...  RYED2SJ8764AF               22.0   
6351          color:Dark Red  ...  RZHEW1JUICN7G  

### Shape e la struttura base

In [58]:
# show the number of rows and columns in the dataset
display('Numero di righe e colonne', dfamazon.shape)

# show the column names in the dataset
display('Nomi delle colonne', dfamazon.columns)

# show the number of non-null values in each column
display(dfamazon.value_counts())



'Numero di righe e colonne'

(6355, 35)

'Nomi delle colonne'

Index(['s_no', 'about_item', 'asin', 'availability', 'best_sellers_rank',
       'brand_name', 'breadcrumbs', 'customer_review_summary',
       'default_variant0', 'default_variant1', 'default_variant2',
       'manufacturer', 'model_number', 'price_value', 'product_description',
       'rating_count', 'rating_distribution1star', 'rating_distribution2star',
       'rating_distribution3star', 'rating_distribution4star',
       'rating_distribution5star', 'rating_stars', 'recent_purchases',
       'seller_name', 'title', 'review_id', 'helpful_vote_count',
       'product_variant', 'rating', 'review_metadata', 'review_position',
       'review_title', 'verified_purchase', 'cleaned_review_text',
       'sentiment_score'],
      dtype='object')

Series([], Name: count, dtype: int64)

Il dataframe in analisi ha 6355 righe e 35 colonne.

Le colonne del dataframe nel dettaglio:
- **s_no**: è un numero intero che contiene gli id di prodotti
- **about_item**: è un oggetto che contiene la descrizione breve del prodotto
- **asin**: è un oggetto che contiene il codice del prodotto
- **availability**: è un oggetto che contiene lo stato del prodotto
- **best_sellers_rank**: è un oggetto che contiene il rating migliore dei consumatori 
- **brand_name**: è un oggetto che contiene il nome del brand
- **breadcrumbs**: è un oggetto che contiene la ierarhia del prodotto ovvero dove si trova
- **customer_review_summary**: è un oggetto che contiene il sommario delle recensioni dei consumatori 
- **default_variant0**: è un oggetto che contiene la taglia/la misura del prodotto
- **default_variant1**: è un oggetto che contiene il colore del prodotto
- **default_variant2**: è un oggetto che contiene le altri descrizioni del prodotto, es. altro colore
- **manufacturer**: è un oggetto che contiene nome dell'azienda che produce il prodotto
- **model_number**: è un oggetto che contiene il numero del modello
- **price_value**: è un numero float che contiene il prezzo corrente
- **product_description**: è un oggetto che contiene la descrizione del prodotto
- **rating_count**: è un oggetto che contiene il numero delle recensioni 
- **rating_distribution1star**: è un oggetto che contiene distribuzione delle stelle(1)
- **rating_distribution2star**: è un oggetto che contiene distribuzione delle stelle(2)
- **rating_distribution3star**: è un oggetto che contiene distribuzione delle stelle(3)
- **rating_distribution4star**: è un oggetto che contiene distribuzione delle stelle(4)
- **rating_distribution5star**: è un oggetto che contiene distribuzione delle stelle(5)
- **rating_stars**: è un oggetto che contiene il rating del prodotto
- **recent_purchases**: è un oggetto che contiene aquisti recenti
- **seller_name**: è un oggetto che contiene il nome del venditore
- **title**: è un oggetto che contiene il titolo
- **review_id**: è un oggetto che contiene id della recensione
- **helpful_vote_count**: è un numero float che contiene numero delle votazioni utili
- **product_variant**: è un oggetto che contiene la variante del prodotto
- **rating**: è un numero float che contiene il rating del prodotto
- **review_metadata**: è un oggetto che contiene i metadati delle recensioni
- **review_position**: è un numero float che contiene la posizione della recensione, come appare sotto il prodotto
- **review_title**: è un oggetto che contiene il titolo della recensione
- **verified_purchase**: è un oggetto che contiene la verificazione dell'aquisto
- **cleaned_review_text**: è un oggetto che contiene il testo della recensione pulito
- **sentiment_score**: è un numero float che contiene il punteggio di sentiment che rappresenta l'orientamento emotivo da -1 a 1(negativo; neutro; positivo)

### Statistiche descrittive

In [59]:
# show names of columns and the data types of each column in the dataset
display('Nomi delle colonne e tipi di dati', dfamazon.info())

# show summary statistics for numerical columns in the dataset
display('Statistiche riassuntive per colonne numeriche', dfamazon.describe())

# show summary statistics for all columns in the dataset
display('Statistiche riassuntive per tutte le colonne', dfamazon.describe(include='all'))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6355 entries, 0 to 6354
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   s_no                      6355 non-null   int64  
 1   about_item                6355 non-null   object 
 2   asin                      6355 non-null   object 
 3   availability              6254 non-null   object 
 4   best_sellers_rank         5001 non-null   object 
 5   brand_name                6355 non-null   object 
 6   breadcrumbs               6288 non-null   object 
 7   customer_review_summary   5829 non-null   object 
 8   default_variant0          6187 non-null   object 
 9   default_variant1          5886 non-null   object 
 10  default_variant2          10 non-null     object 
 11  manufacturer              2369 non-null   object 
 12  model_number              1769 non-null   object 
 13  price_value               6355 non-null   float64
 14  product_

'Nomi delle colonne e tipi di dati'

None

'Statistiche riassuntive per colonne numeriche'

s_no  price_value  helpful_vote_count       rating  \
count  6355.000000  6355.000000         6327.000000  6320.000000   
mean    411.259481    33.705065            3.336336     4.533228   
std     290.009883    23.617557           19.090762     0.855710   
min       0.000000     0.000000            0.000000     1.000000   
25%     175.000000    19.900000            0.000000     4.000000   
50%     357.000000    28.410000            0.000000     5.000000   
75%     542.500000    40.010000            2.000000     5.000000   
max    1011.000000   249.990000          710.000000     5.000000   

       review_position  sentiment_score  
count      6327.000000      6327.000000  
mean          5.374743         0.307546  
std           2.889921         0.218020  
min           1.000000        -1.000000  
25%           3.000000         0.168600  
50%           5.000000         0.300000  
75%           8.000000         0.450000  
max          10.000000         1.000000

'Statistiche riassuntive per tutte le colonne'

s_no                                         about_item  \
count   6355.000000                                               6355   
unique          NaN                                                697   
top             NaN  COMFORT FLEX WAISTBAND: Constructed with comfo...   
freq            NaN                                                 33   
mean     411.259481                                                NaN   
std      290.009883                                                NaN   
min        0.000000                                                NaN   
25%      175.000000                                                NaN   
50%      357.000000                                                NaN   
75%      542.500000                                                NaN   
max     1011.000000                                                NaN   

              asin availability  \
count         6355         6254   
unique         728           11   
top     B074KL8RVS     In Stock   
freq            19         5739   
mean           NaN          NaN   
std            NaN          NaN   
min            NaN          NaN   
25%            NaN          NaN   
50%            NaN          NaN   
75%            NaN          NaN   
max            NaN          NaN   

                                        best_sellers_rank   brand_name  \
count                                                5001         6355   
unique                                                539          285   
top     #470 in Clothing, Shoes & Jewelry (See Top 100...  Hanes Store   
freq                                                   33          319   
mean                                                  NaN          NaN   
std                                                   NaN          NaN   
min                                                   NaN          NaN   
25%                                                   NaN          NaN   
50%                                                   NaN          NaN   
75%                                                   NaN          NaN   
max                                                   NaN          NaN   

                                              breadcrumbs  \
count                                                6288   
unique                                                140   
top     Clothing, Shoes & Jewelry › Men › Clothing › J...   
freq                                                  511   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                  customer_review_summary default_variant0  \
count                                                5829             6187   
unique                                                616              118   
top     Customers find the shirts comfortable and soft...       size:Large   
freq                                                   14             1528   
mean                                                  NaN              NaN   
std                                                   NaN              NaN   
min                                                   NaN              NaN   
25%                                                   NaN              NaN   
50%                                                   NaN              NaN   
75%                                                   NaN              NaN   
max                                                   NaN              NaN   

       default_variant1  ...       review_id helpful_vote_count  \
count              5886  ...            6327        6327.000000   
unique      

### Valori mancanti (Missing Values) e Duplicati

#### Dati mancanti isnull()

In [60]:
# create a variable to count missing values in each column
valuesmissing = dfamazon.isnull().sum()
# control missing values in each column taking notnull values and sort them in descending order
valuesmissing[valuesmissing > 0].sort_values(ascending=False)


default_variant2           6345
model_number               4586
product_description        4000
manufacturer               3986
best_sellers_rank          1354
recent_purchases           1098
customer_review_summary     526
product_variant             519
default_variant1            469
default_variant0            168
seller_name                 107
availability                101
breadcrumbs                  67
cleaned_review_text          67
review_metadata              59
review_title                 42
rating                       35
review_position              28
helpful_vote_count           28
review_id                    28
verified_purchase            28
sentiment_score              28
rating_count                  9
rating_stars                  9
dtype: int64

In [61]:
# control missing values in each column taking notnull values, sort them in descending order 
# and show percentage of missing values
(valuesmissing[valuesmissing > 0].sort_values(ascending=False) / len(dfamazon) * 100).round(3)

default_variant2           99.843
model_number               72.164
product_description        62.943
manufacturer               62.722
best_sellers_rank          21.306
recent_purchases           17.278
customer_review_summary     8.277
product_variant             8.167
default_variant1            7.380
default_variant0            2.644
seller_name                 1.684
availability                1.589
breadcrumbs                 1.054
cleaned_review_text         1.054
review_metadata             0.928
review_title                0.661
rating                      0.551
review_position             0.441
helpful_vote_count          0.441
review_id                   0.441
verified_purchase           0.441
sentiment_score             0.441
rating_count                0.142
rating_stars                0.142
dtype: float64

In [62]:
# create DF pd.DataFrame
dfamazon_missing = pd.DataFrame(
    {'Colonna': dfamazon.columns,
     'N_Dati_mancanti': (dfamazon.isnull().sum()).round(2),
     '%_Dati_mancanti': (dfamazon.isnull().sum() / len(dfamazon) * 100).round(2),
    }
).query('N_Dati_mancanti > 0').sort_values('N_Dati_mancanti', ascending=False)
display(dfamazon_missing)

Colonna  N_Dati_mancanti  \
default_variant2                default_variant2             6345   
model_number                        model_number             4586   
product_description          product_description             4000   
manufacturer                        manufacturer             3986   
best_sellers_rank              best_sellers_rank             1354   
recent_purchases                recent_purchases             1098   
customer_review_summary  customer_review_summary              526   
product_variant                  product_variant              519   
default_variant1                default_variant1              469   
default_variant0                default_variant0              168   
seller_name                          seller_name              107   
availability                        availability              101   
breadcrumbs                          breadcrumbs               67   
cleaned_review_text          cleaned_review_text               67   
review_metadata                  review_metadata               59   
review_title                        review_title               42   
rating                                    rating               35   
review_position                  review_position               28   
helpful_vote_count            helpful_vote_count               28   
review_id                              review_id               28   
verified_purchase              verified_purchase               28   
sentiment_score                  sentiment_score               28   
rating_count                        rating_count                9   
rating_stars                        rating_stars                9   

                         %_Dati_mancanti  
default_variant2                   99.84  
model_number                       72.16  
product_description                62.94  
manufacturer                       62.72  
best_sellers_rank                  21.31  
recent_purchases                   17.28  
customer_review_summary             8.28  
product_variant                     8.17  
default_variant1                    7.38  
default_variant0                    2.64  
seller_name                         1.68  
availability                        1.59  
breadcrumbs                         1.05  
cleaned_review_text                 1.05  
review_metadata                     0.93  
review_title                        0.66  
rating                              0.55  
review_position                     0.44  
helpful_vote_count                  0.44  
review_id                           0.44  
verified_purchase                   0.44  
sentiment_score                     0.44  
rating_count                        0.14  
rating_stars                        0.14

#### Dati duplicati (duplicated())

In [63]:
# check for duplicate rows in the dataset
dfamazon.duplicated().sum()
# show duplicate rows based on 'asin' and 'review_id' columns
dfamazon[dfamazon.duplicated(subset=['asin','review_id'], keep=False)]

Empty DataFrame
Columns: [s_no, about_item, asin, availability, best_sellers_rank, brand_name, breadcrumbs, customer_review_summary, default_variant0, default_variant1, default_variant2, manufacturer, model_number, price_value, product_description, rating_count, rating_distribution1star, rating_distribution2star, rating_distribution3star, rating_distribution4star, rating_distribution5star, rating_stars, recent_purchases, seller_name, title, review_id, helpful_vote_count, product_variant, rating, review_metadata, review_position, review_title, verified_purchase, cleaned_review_text, sentiment_score]
Index: []

[0 rows x 35 columns]

#### Cancellazione dei dati mancanti e duplicati nel dataset

In [64]:
# create a copy of the original dataset
dfamazoncopy = dfamazon.copy()
dfamazoncopy.head()

s_no                                         about_item        asin  \
0   146  SLIM FIT. Designed with a functional fit in mi...  B0006U6C0Q   
1    54  DESIGNS: Right sleeve with a pocket of the pol...  B09TK9248H   
2   103  Super-light HeatGear fabric delivers superior ...  B0874WD2B2   
3    56  100% Polyester, Quick Dry & Lightweight，Camo p...  B0C1SC8PR9   
4    43  High Performance Material: Polo shirts for men...  B0D38YFY3K   

  availability                                  best_sellers_rank  \
0     In Stock  #1,493 in Clothing, Shoes & Jewelry (See Top 1...   
1     In Stock  #49,564 in Clothing, Shoes & Jewelry (See Top ...   
2     In Stock  #662 in Clothing, Shoes & Jewelry (See Top 100...   
3     In Stock  #44,945 in Clothing, Shoes & Jewelry (See Top ...   
4     In Stock  #32,242 in Clothing, Shoes & Jewelry (See Top ...   

           brand_name                                        breadcrumbs  \
0      Wrangler Store  Clothing, Shoes & Jewelry › Men › Clothing › J...   
1          ZITY Store  Clothing, Shoes & Jewelry › Men › Clothing › A...   
2  Under Armour Store  Clothing, Shoes & Jewelry › Men › Clothing › A...   
3         MoFiz Store  Clothing, Shoes & Jewelry › Men › Clothing › S...   
4       ZENGVEE Store  Clothing, Shoes & Jewelry › Men › Clothing › A...   

                             customer_review_summary default_variant0  \
0  Customers appreciate the jeans for their quali...   size:34W x 29L   
1  Customers appreciate the nice colors and comfo...     size:X-Large   
2  Customers find that the compression shirts fit...      size:Medium   
3  Customers appreciate the comfortable fit, nice...       size:Large   
4  Customers appreciate the shirts' fit, look, an...       size:Large   

                     default_variant1  ...       review_id helpful_vote_count  \
0              color:Prewashed Indigo  ...  R103VSGVPRSZ6F              230.0   
1                         color:Gnblg  ...  R103YA89AQKKT4                0.0   
2                                 NaN  ...  R105XJ6QWC0GUL                0.0   
3             color:B-army Green Camo  ...  R1084B47TWDC0Z                0.0   
4  color:1black Gray Green Navy White  ...  R10A6DXDSXNG8V                0.0   

                                     product_variant  rating  \
0             Color: Prewashed IndigoSize: 32W x 34L     5.0   
1  Size: X-LargeColor: GnblkHeight: 6'3Weight: 21...     5.0   
2                Size: SmallColor: White (100)/Black     5.0   
3  Size: X-LargeColor: B-army Green CamoHeight: 5...     5.0   
4   Size: X-LargeColor: 1black Gray Green Blue Khaki     4.0   

                                     review_metadata review_position  \
0  Reviewed in the United States on September 18,...             2.0   
1  Reviewed in the United States on December 16, ...             4.0   
2  Reviewed in the United States on January 18, 2025             8.0   
3  Reviewed in the United States on December 26, ...             4.0   
4  Reviewed in the United States on November 11, ...             8.0   

                         review_title verified_purchase  \
0               For the 936PWD Style!              True   
1                You should buy these              True   
2   Keep a you cool while working out              True   
3     Very comfortable and breathable              True   
4  Good but a skosh on the small side              True   

                                 cleaned_review_text sentiment_score  
0  husband wrangler man probably last year used h...          0.1340  
1  low expectation cheap going give son turn husb...          0.3563  
2  got wear hockey equipment keep cool dry run li...          0.1292  
3  liked true size good price lightweight comfort...          0.5100  
4  comfortable well made tad bit toward small sid...          0.2833  

[5 rows x 35 columns]

In [65]:
# delete the trash columns with missing values > 60%
cols_trash = dfamazoncopy.columns[dfamazoncopy.isnull().mean() > 0.6].tolist()
print(cols_trash)

['default_variant2', 'manufacturer', 'model_number', 'product_description']


In [66]:
# drop trash columns from the copied dataset
dfamazoncopy = dfamazoncopy.drop(columns=cols_trash)
dfamazoncopy.head(3)
dfamazoncopy.shape

(6355, 31)

#### Controllo il DataFrame dopo la pulizia

In [67]:
dfamazoncopy.isnull().mean()

s_no                        0.000000
about_item                  0.000000
asin                        0.000000
availability                0.015893
best_sellers_rank           0.213061
brand_name                  0.000000
breadcrumbs                 0.010543
customer_review_summary     0.082769
default_variant0            0.026436
default_variant1            0.073800
price_value                 0.000000
rating_count                0.001416
rating_distribution1star    0.000000
rating_distribution2star    0.000000
rating_distribution3star    0.000000
rating_distribution4star    0.000000
rating_distribution5star    0.000000
rating_stars                0.001416
recent_purchases            0.172777
seller_name                 0.016837
title                       0.000000
review_id                   0.004406
helpful_vote_count          0.004406
product_variant             0.081668
rating                      0.005507
review_metadata             0.009284
review_position             0.004406
r

In [68]:
# create DF pd.DataFrame
# update to dfamazoncopy
# show updated missing values after dropping trash columns
dfamazon_missing = pd.DataFrame(
    {'Colonna': dfamazoncopy.columns,
     'N_Dati_mancanti': (dfamazoncopy.isnull().sum()).round(2),
     '%_Dati_mancanti': (dfamazoncopy.isnull().sum() / len(dfamazon) * 100).round(2),
    }
).query('N_Dati_mancanti > 0').sort_values('N_Dati_mancanti', ascending=False)
display(dfamazon_missing)

Colonna  N_Dati_mancanti  \
best_sellers_rank              best_sellers_rank             1354   
recent_purchases                recent_purchases             1098   
customer_review_summary  customer_review_summary              526   
product_variant                  product_variant              519   
default_variant1                default_variant1              469   
default_variant0                default_variant0              168   
seller_name                          seller_name              107   
availability                        availability              101   
breadcrumbs                          breadcrumbs               67   
cleaned_review_text          cleaned_review_text               67   
review_metadata                  review_metadata               59   
review_title                        review_title               42   
rating                                    rating               35   
review_position                  review_position               28   
helpful_vote_count            helpful_vote_count               28   
review_id                              review_id               28   
verified_purchase              verified_purchase               28   
sentiment_score                  sentiment_score               28   
rating_count                        rating_count                9   
rating_stars                        rating_stars                9   

                         %_Dati_mancanti  
best_sellers_rank                  21.31  
recent_purchases                   17.28  
customer_review_summary             8.28  
product_variant                     8.17  
default_variant1                    7.38  
default_variant0                    2.64  
seller_name                         1.68  
availability                        1.59  
breadcrumbs                         1.05  
cleaned_review_text                 1.05  
review_metadata                     0.93  
review_title                        0.66  
rating                              0.55  
review_position                     0.44  
helpful_vote_count                  0.44  
review_id                           0.44  
verified_purchase                   0.44  
sentiment_score                     0.44  
rating_count                        0.14  
rating_stars                        0.14

### Creazione nuove colonne utili per l'analisi

#### Creo le colonne categoria

In [69]:
#dfamazoncopy['breadcrumbs'].head(10)
# split 'breadcrumbs' column into multiple columns
dfamazoncopy['category1'] = dfamazoncopy['breadcrumbs'].str.split(' › ').str[0].fillna('')
dfamazoncopy['category2'] = dfamazoncopy['breadcrumbs'].str.split(' › ').str[1].fillna('')
dfamazoncopy['category3'] = dfamazoncopy['breadcrumbs'].str.split(' › ').str[2].fillna('')
dfamazoncopy['category4'] = dfamazoncopy['breadcrumbs'].str.split(' › ').str[3].fillna('')
dfamazoncopy['category5'] = dfamazoncopy['breadcrumbs'].str.split(' › ').str[4].fillna('')
dfamazoncopy['category6'] = dfamazoncopy['breadcrumbs'].str.split(' › ').str[5].fillna('')

# drop original 'breadcrumbs' column
dfamazoncopy.drop(['breadcrumbs'], axis=1, inplace=True)
dfamazoncopy.head(10)

display(dfamazoncopy['verified_purchase'].dtype)

dtype('O')

#### Convertire valori in colonna

In [70]:
def safe_to_bool(series):
    #Converte stringhe in boolean senza warning#
    mapping = {'true': True, 'false': False, '1': True, '0': False}
    return (series.astype(str).str.lower()
            .str.strip()
            .map(mapping)
            .fillna(False)
            .astype(bool))

dfamazoncopy['verified_purchase'] = safe_to_bool(dfamazoncopy['verified_purchase'])

display(dfamazoncopy['verified_purchase'].dtype)
display(dfamazoncopy['verified_purchase'].value_counts())

C:\Users\39347\AppData\Local\Temp\ipykernel_17308\3193357457.py:7: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .fillna(False)


dtype('bool')

verified_purchase
True     6167
False     188
Name: count, dtype: int64

#### Availability

In [71]:
mapping_availability = {
    'in stock': True,
    'only 1 left in stock - order soon.': True,
    'only 3 left in stock - order soon.': True,
    'currently unavailable.': False,
    'only 2 left in stock - order soon.': True,
    'only 5 left in stock - order soon.': True,
    'available to ship in 1-2 days': True,
    'this item will be released on may 20, 2025.': True,
    'only 4 left in stock - order soon.': True,
    'temporarily out of stock.': False,
}

def map_availability(text):
    text = str(text).lower().strip()
    # default False se la stringa non è nelle chiavi
    return mapping_availability.get(text, False)

dfamazoncopy['availability_b'] = (
    dfamazoncopy['availability']
      .apply(map_availability)
      .astype(bool)
)

print(dfamazoncopy['availability_b'].value_counts())
#cancello la colonna availability
dfamazoncopy.drop(['availability'],axis=1, inplace=True)
#dfamazoncopy.head(3)

availability_b
True     6166
False     189
Name: count, dtype: int64


#### Controllo colonna rating_count

In [72]:
#print(dfamazoncopy['rating_count'].value_counts())
dfamazoncopy['rating_count_n'] = (
    dfamazoncopy['rating_count']
    .astype(str)
    .str.replace('ratings','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
    .str.replace(',','', regex=False) # rimuove le virgole nei numeri grandi
)
# convert to numeric
dfamazoncopy['rating_count_n'] = pd.to_numeric(
    dfamazoncopy['rating_count_n'], 
    errors='coerce' # convert errors to NaN
).astype('Int64') # usa Int64 per supportare NaN

dfamazoncopy.drop(['rating_count'], axis=1, inplace=True)

print(dfamazoncopy['rating_count_n'].value_counts())    

rating_count_n
39       30
13       26
26       24
14       22
22       22
         ..
1058      1
26060     1
6477      1
247       1
8110      1
Name: count, Length: 644, dtype: Int64


#### Colonna rating_stars


In [73]:
#dfamazoncopy['rating_stars'].value_counts()

dfamazoncopy['rating_stars_n'] = (
    dfamazoncopy['rating_stars']
    .astype(str)
    .str.replace('out of 5 stars','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
)
# convert to numeric
dfamazoncopy['rating_stars_n'] = pd.to_numeric(
    dfamazoncopy['rating_stars_n'], 
    errors='coerce' # convert errors to NaN
).astype('Float64') # usa Int64 per supportare NaN

dfamazoncopy.drop(['rating_stars'], axis=1, inplace=True)

print(dfamazoncopy['rating_stars_n'].value_counts()) 

rating_stars_n
4.5    1213
4.6    1146
4.4    1052
4.3     825
4.2     604
4.7     440
4.8     317
4.1     241
4.0     158
3.8      96
3.9      67
3.6      50
3.7      50
3.4      14
3.3      12
5.0      10
3.2      10
4.9      10
3.1       9
3.0       9
3.5       9
2.2       2
1.0       1
2.6       1
Name: count, dtype: Int64


#### Colonna recent_purchases

In [74]:
dfamazoncopy['recent_purchases'].value_counts()

dfamazoncopy['recent_purchases_n'] = (
    dfamazoncopy['recent_purchases']
    .astype(str)
    .str.replace('+ bought','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
    .str.replace('K','000', regex=False) # rimuove le virgole nei numeri grandi
)
# convert to numeric
dfamazoncopy['recent_purchases_n'] = pd.to_numeric(
    dfamazoncopy['recent_purchases_n'], 
    errors='coerce' # convert errors to NaN
).astype('Int64') # usa Int64 per supportare NaN

dfamazoncopy.drop(['recent_purchases'], axis=1, inplace=True)

print(dfamazoncopy['recent_purchases_n'].value_counts())  

recent_purchases_n
100      1008
200       704
1000      579
300       563
400       508
50        478
500       329
2000      223
600       213
700       182
800       142
3000      106
900        84
4000       62
5000       33
9000       12
10000      11
8000       10
7000       10
Name: count, dtype: Int64


#### Colonna rating_distribuition1star

In [75]:
#dfamazoncopy['rating_distribution1star'].value_counts()
dfamazoncopy['rating_distribution1star_%'] = (
    dfamazoncopy['rating_distribution1star']
    .astype(str)
    .str.replace('%','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
)

# convert to numeric
dfamazoncopy['rating_distribution1star_%'] = pd.to_numeric(
    dfamazoncopy['rating_distribution1star_%'], 
    errors='coerce' # convert errors to NaN
).astype('Float64') # usa Int64 per supportare NaN

# convert percentage to fraction
dfamazoncopy['rating_distribution1star_%'] = (
    dfamazoncopy['rating_distribution1star_%']
    .fillna(0)
    .div(100)
)

dfamazoncopy.drop(['rating_distribution1star'], axis=1, inplace=True)
dfamazoncopy['rating_distribution1star_%'].value_counts()


rating_distribution1star_%
0.03    1107
0.04    1039
0.02     928
0.05     800
0.06     511
0.07     408
0.0      316
0.01     303
0.08     221
0.09     208
0.1      112
0.12      66
0.16      60
0.11      56
0.14      51
0.15      38
0.13      30
0.18      26
0.19      11
0.28      10
0.24      10
0.26       9
0.35       8
0.3        5
0.27       5
0.39       4
0.2        3
0.17       2
0.47       2
0.23       2
0.46       1
1.0        1
0.48       1
0.32       1
Name: count, dtype: Int64

In [76]:
dfamazoncopy['rating_distribution2star_%'] = (
    dfamazoncopy['rating_distribution2star']
    .astype(str)
    .str.replace('%','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
)

# convert to numeric
dfamazoncopy['rating_distribution2star_%'] = pd.to_numeric(
    dfamazoncopy['rating_distribution2star_%'], 
    errors='coerce' # convert errors to NaN
).astype('Float64') # usa Int64 per supportare NaN

# convert percentage to fraction
dfamazoncopy['rating_distribution2star_%'] = (
    dfamazoncopy['rating_distribution2star_%']
    .fillna(0)
    .div(100)
)

dfamazoncopy.drop(['rating_distribution2star'], axis=1, inplace=True)
dfamazoncopy['rating_distribution2star_%'].value_counts()

rating_distribution2star_%
0.02    1766
0.03    1367
0.01    1003
0.04     810
0.0      553
0.05     388
0.06     196
0.07     113
0.08      61
0.09      41
0.21      12
0.1       11
0.13      10
0.12       8
0.14       7
0.18       5
0.3        2
0.22       1
0.17       1
Name: count, dtype: Int64

In [77]:
dfamazoncopy['rating_distribution3star_%'] = (
    dfamazoncopy['rating_distribution3star']
    .astype(str)
    .str.replace('%','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
)

# convert to numeric
dfamazoncopy['rating_distribution3star_%'] = pd.to_numeric(
    dfamazoncopy['rating_distribution3star_%'], 
    errors='coerce' # convert errors to NaN
).astype('Float64') # usa Int64 per supportare NaN

# convert percentage to fraction
dfamazoncopy['rating_distribution3star_%'] = (
    dfamazoncopy['rating_distribution3star_%']
    .fillna(0)
    .div(100)
)

dfamazoncopy.drop(['rating_distribution3star'], axis=1, inplace=True)
dfamazoncopy['rating_distribution3star_%'].value_counts()

rating_distribution3star_%
0.06    976
0.05    925
0.08    826
0.07    788
0.09    565
0.04    495
0.1     450
0.03    443
0.11    212
0.0     141
0.12    136
0.02    125
0.13    101
0.15     30
0.14     25
0.2      20
0.18     20
0.01     20
0.16     19
0.19     13
0.17     13
0.23      7
0.21      3
0.25      2
Name: count, dtype: Int64

In [78]:
dfamazoncopy['rating_distribution4star_%'] = (
    dfamazoncopy['rating_distribution4star']
    .astype(str)
    .str.replace('%','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
)

# convert to numeric
dfamazoncopy['rating_distribution4star_%'] = pd.to_numeric(
    dfamazoncopy['rating_distribution4star_%'], 
    errors='coerce' # convert errors to NaN
).astype('Float64') # usa Int64 per supportare NaN

# convert percentage to fraction
dfamazoncopy['rating_distribution4star_%'] = (
    dfamazoncopy['rating_distribution4star_%']
    .fillna(0)
    .div(100)
)

dfamazoncopy.drop(['rating_distribution4star'], axis=1, inplace=True)
dfamazoncopy['rating_distribution4star_%'].value_counts()

rating_distribution4star_%
0.16    850
0.15    767
0.17    766
0.14    721
0.13    597
0.18    468
0.12    414
0.11    305
0.1     254
0.09    241
0.19    199
0.08    155
0.2     150
0.23     84
0.22     45
0.21     44
0.0      39
0.07     37
0.26     32
0.24     31
0.25     25
0.28     24
0.29     21
0.06     21
0.3      16
0.31     11
0.32     10
0.05     10
0.37      6
0.41      5
0.34      4
0.75      2
0.54      1
Name: count, dtype: Int64

In [79]:
dfamazoncopy['rating_distribution5star_%'] = (
    dfamazoncopy['rating_distribution5star']
    .astype(str)
    .str.replace('%','', regex=False)
    .str.strip() # rimuove spazi bianchi iniziali e finali
)

# convert to numeric
dfamazoncopy['rating_distribution5star_%'] = pd.to_numeric(
    dfamazoncopy['rating_distribution5star_%'], 
    errors='coerce' # convert errors to NaN
).astype('Float64') # usa Int64 per supportare NaN

# convert percentage to fraction
dfamazoncopy['rating_distribution5star_%'] = (
    dfamazoncopy['rating_distribution5star_%']
    .fillna(0)
    .div(100)
)

dfamazoncopy.drop(['rating_distribution5star'], axis=1, inplace=True)
dfamazoncopy['rating_distribution5star_%'].value_counts()

rating_distribution5star_%
0.75    409
0.74    357
0.73    326
0.66    318
0.68    297
       ... 
0.43      5
0.47      5
0.27      4
0.38      2
0.23      2
Name: count, Length: 63, dtype: Int64

In [80]:
dfamazoncopy.head(20)

s_no                                         about_item        asin  \
0    146  SLIM FIT. Designed with a functional fit in mi...  B0006U6C0Q   
1     54  DESIGNS: Right sleeve with a pocket of the pol...  B09TK9248H   
2    103  Super-light HeatGear fabric delivers superior ...  B0874WD2B2   
3     56  100% Polyester, Quick Dry & Lightweight，Camo p...  B0C1SC8PR9   
4     43  High Performance Material: Polo shirts for men...  B0D38YFY3K   
5    103  Super-light HeatGear fabric delivers superior ...  B0874WD2B2   
6    153  Slim fit,simple style,basic jeans you need eve...  B01N58LDY4   
7     76  America's #1 Men's Underwear Brand - Source: C...  B086L1PM8V   
8     87  Named Best Anti-sweat undershirt by Men’s Heal...  B074XY7MTG   
9      9  Performance Fabric Blend: material blend of 95...  B0CXDL2NP6   
10    50  Crafted with advanced breathable fabric,Our Ac...  B0DR7WSCFQ   
11     9  Performance Fabric Blend: material blend of 95...  B0CXDL2NP6   
12   155  Same great jeans, brand new name! Signature by...  B073V3JSSB   
13    28     100% CottonImportedPull On closureMachine Wash  B01AZ4DAJ8   
14    55  [Feature]:These mens polo shirts with collar c...  B0D9GXG945   
15    16  Soft Breathable FabricLay Flat CollarWicks Awa...  B0B47Q6QYP   
16   138  RELAXED FIT. These jeans feature a roomier fit...  B09NMN9C21   
17    73  THE FIT - Standard-fit men's t-shirt.THE FEEL ...  B078LC7GCX   
18   114  Regular fit thru seat and thigh5 pocket stylin...  B07LGQ884Z   
19    75  Rib-knit crew neckSide-seam construction minim...  B002GHC19I   

                                    best_sellers_rank  \
0   #1,493 in Clothing, Shoes & Jewelry (See Top 1...   
1   #49,564 in Clothing, Shoes & Jewelry (See Top ...   
2   #662 in Clothing, Shoes & Jewelry (See Top 100...   
3   #44,945 in Clothing, Shoes & Jewelry (See Top ...   
4   #32,242 in Clothing, Shoes & Jewelry (See Top ...   
5   #662 in Clothing, Shoes & Jewelry (See Top 100...   
6   #13,904 in Clothing, Shoes & Jewelry (See Top ...   
7   #580 in Clothing, Shoes & Jewelry (See Top 100...   
8   #1,936 in Clothing, Shoes & Jewelry (See Top 1...   
9   #5,623 in Clothing, Shoes & Jewelry (See Top 1...   
10  #10,859 in Clothing, Shoes & Jewelry (See Top ...   
11  #5,623 in Clothing, Shoes & Jewelry (See Top 1...   
12  #3,604 in Clothing, Shoes & Jewelry (See Top 1...   
13  #88,984 in Clothing, Shoes & Jewelry (See Top ...   
14  #13,681 in Clothing, Shoes & Jewelry (See Top ...   
15  #47,659 in Clothing, Shoes & Jewelry (See Top ...   
16  #2,429 in Clothing, Shoes & Jewelry (See Top 1...   
17  #395 in Clothing, Shoes & Jewelry (See Top 100...   
18  #78 in Clothing, Shoes & Jewelry (See Top 100 ...   
19  #18 in Clothing, Shoes & Jewelry (See Top 100 ...   

                           brand_name  \
0                      Wrangler Store   
1                          ZITY Store   
2                  Under Armour Store   
3                         MoFiz Store   
4                       ZENGVEE Store   
5                  Under Armour Store   
6                FREDD MARSHALL Store   
7                         Hanes Store   
8                         Hanes Store   
9                       OYGSieg Store   
10                    Brand: KAOKLRNI   
11                      OYGSieg Store   
12  Levi Strauss Signature Gold Store   
13             Brand: U.S. Polo Assn.   
14                       Qekoma Store   
15            POLO RALPH LAUREN Store   
16                     Wrangler Store   
17                     Champion Store   
18                      Rustler Store   
19                     Carhartt Store   

                              customer_review_summary default_variant0  \
0   Customers appreciate the jeans for their quali...   size:34W x 29L   
1   Customers appreciate the nice colors and comfo...     size:X-Large   
2   Customers find that the compression shirts fit...      size:Medium   
3   Customers appreciate the comfortable fit, nice...       size:Large  

### Tabelle di aggregazione e Group By

#### Tabella di aggregazione per Recensioni per prodotto

andando a mostrare i top 10 prodotti che hanno avuto più recensioni

In [81]:
# table
product_reviews = dfamazoncopy.groupby('title').agg({
    'review_id':'count', # number of reviews per product
    'rating' :'mean', # mean rating 
    'helpful_vote_count': 'mean' # mean helpful votes
}).round(2)

# rename columns
product_reviews.columns=['N_Review', 'Mean_Rating', 'Mean_Helpful_Votes']
product_reviews.sort_values('N_Review', ascending=False).head(10)

N_Review  Mean_Rating  \
title                                                                       
Lee Men's Regular Fit Straight Leg Jean                   20         4.70   
Hanes Men's EcoSmart Fleece, Pullover Crewneck ...        20         4.60   
Wrangler Men's Regular Fit Comfort Flex Waist Jean        19         4.74   
Wrangler Authentics Men's Regular Fit Comfort F...        14         4.93   
Nautica Men's Short Sleeve Solid Crew Neck T-Shirt        14         4.64   
Real Essentials 5 Pack: Men’s Short Sleeve Dry ...        14         4.29   
Dockers Men's Perfect Classic Fit 8 Shorts                13         4.38   
Gildan Men's Crew T-Shirts, Multipack, Style G1100        13         4.62   
Workout Shirts for Men Short Sleeve Quick Dry A...        12         4.58   
Lee Men's Extreme Motion Flat Front Short                 12         4.75   

                                                    Mean_Helpful_Votes  
title                                                                   
Lee Men's Regular Fit Straight Leg Jean                           0.35  
Hanes Men's EcoSmart Fleece, Pullover Crewneck ...               12.15  
Wrangler Men's Regular Fit Comfort Flex Waist Jean                0.63  
Wrangler Authentics Men's Regular Fit Comfort F...                5.21  
Nautica Men's Short Sleeve Solid Crew Neck T-Shirt                0.29  
Real Essentials 5 Pack: Men’s Short Sleeve Dry ...                5.29  
Dockers Men's Perfect Classic Fit 8 Shorts                        0.69  
Gildan Men's Crew T-Shirts, Multipack, Style G1100                6.00  
Workout Shirts for Men Short Sleeve Quick Dry A...                5.17  
Lee Men's Extreme Motion Flat Front Short                         1.17

In [82]:
dfamazoncopy.columns

Index(['s_no', 'about_item', 'asin', 'best_sellers_rank', 'brand_name',
       'customer_review_summary', 'default_variant0', 'default_variant1',
       'price_value', 'seller_name', 'title', 'review_id',
       'helpful_vote_count', 'product_variant', 'rating', 'review_metadata',
       'review_position', 'review_title', 'verified_purchase',
       'cleaned_review_text', 'sentiment_score', 'category1', 'category2',
       'category3', 'category4', 'category5', 'category6', 'availability_b',
       'rating_count_n', 'rating_stars_n', 'recent_purchases_n',
       'rating_distribution1star_%', 'rating_distribution2star_%',
       'rating_distribution3star_%', 'rating_distribution4star_%',
       'rating_distribution5star_%'],
      dtype='object')

#### Tabella di aggregazione per il voto medio per categoria

In [85]:
# table rating_stars mean
cat_rating = dfamazoncopy.groupby(['category1', 'category2', 'category3'])['rating'].agg(['mean','count']).round(2)
# rename columns
cat_rating.columns=['Mean_Rating_Stars', 'N_reviews']
cat_rating.sort_values('Mean_Rating_Stars', ascending=False)

Mean_Rating_Stars  \
category1                 category2               category3                                        
Clothing, Shoes & Jewelry Baby                    Baby Boys                                 4.83   
                                                  Baby Girls                                4.83   
                          Sport Specific Clothing Skiing                                    4.80   
                          Men                     Shoes                                     4.68   
                          Boys                    Clothing                                  4.67   
                          Women                   Accessories                               4.65   
                          Sport Specific Clothing Running                                   4.63   
                          Men                     Uniforms, Work & Safety                   4.62   
                          Sport Specific Clothing Hiking & Outdoor Recreation               4.60   
Sports & Outdoors         Clothing                Men                                       4.60   
Clothing, Shoes & Jewelry Sport Specific Clothing Golf                                      4.58   
                          Men                     Clothing                                  4.54   
                          Women                   Shoes                                     4.54   
                          Men                                                               4.50   
                          Women                   Clothing                                  4.44   
                          Sport Specific Clothing Yoga                                      4.20   
Electronics               Computers & Accessories Laptop Accessories                        4.00   
Clothing, Shoes & Jewelry Women                                                             3.93   
                          Costumes & Accessories  Men                                       3.90   
                          Novelty & More          Clothing                                  3.83   
                                                                                            3.44   
Clothing, Shoes & Jewelry Novelty & More                                                    3.00   
                          Boys                                                               NaN   

                                                                               N_reviews  
category1                 category2               category3                               
Clothing, Shoes & Jewelry Baby                    Baby Boys                          110  
                                                  Baby Girls                         201  
                          Sport Specific Clothing Skiing                              70  
                          Men                     Shoes                              423  
                          Boys                    Clothing                           296  
                          Women                   Accessories                         20  
                          Sport Specific Clothing Running                             30  
                          Men                     Uniforms, Work & Safety            112  
                          Sport Specific Clothing Hiking & Outdoor Recreation         10  
Sports & Outdoors         Clothing                Men                                 10  
Clothing, Shoes & Jewelry Sport Specific Clothing Golf                                33  
                          Men                     Clothing                          3392  
                          Women                   Shoes                              325  
                          Men                                                         10  
                          Women                   Clothing                          1078  
                          Sport

#### Pivot: Rating per brand-categoria

In [88]:
pivot_rating = dfamazoncopy.pivot_table(
    values='rating',
    index=['category1', 'category2'],
    columns=['category3'],
    aggfunc=['mean', 'count'],
    fill_value=0
).round(2)
pivot_rating.head(10)

mean                        \
category3                                               Accessories Baby Boys   
category1                 category2                                             
                                                   3.44        0.00      0.00   
Clothing, Shoes & Jewelry Baby                     0.00        0.00      4.83   
                          Boys                     0.00        0.00      0.00   
                          Costumes & Accessories   0.00        0.00      0.00   
                          Men                      4.50        0.00      0.00   
                          Novelty & More           3.00        0.00      0.00   
                          Sport Specific Clothing  0.00        0.00      0.00   
                          Women                    3.93        4.65      0.00   
Electronics               Computers & Accessories  0.00        0.00      0.00   
Sports & Outdoors         Clothing                 0.00        0.00      0.00   

                                                                             \
category3                                         Baby Girls Clothing  Golf   
category1                 category2                                           
                                                        0.00     0.00  0.00   
Clothing, Shoes & Jewelry Baby                          4.83     0.00  0.00   
                          Boys                          0.00     4.67  0.00   
                          Costumes & Accessories        0.00     0.00  0.00   
                          Men                           0.00     4.54  0.00   
                          Novelty & More                0.00     3.83  0.00   
                          Sport Specific Clothing       0.00     0.00  4.58   
                          Women                         0.00     4.44  0.00   
Electronics               Computers & Accessories       0.00     0.00  0.00   
Sports & Outdoors         Clothing                      0.00     0.00  0.00   

                                                                               \
category3                                         Hiking & Outdoor Recreation   
category1                 category2                                             
                                                                          0.0   
Clothing, Shoes & Jewelry Baby                                            0.0   
                          Boys                                            0.0   
                          Costumes & Accessories                          0.0   
                          Men                                             0.0   
                          Novelty & More                                  0.0   
                          Sport Specific Clothing                         4.6   
                          Women                                           0.0   
Electronics               Computers & Accessories                         0.0   
Sports & Outdoors         Clothing                                        0.0   

                                                                           \
category3                                         Laptop Accessories  Men   
category1                 category2                                         
                                                                 0.0  0.0   
Clothing, Shoes & Jewelry Baby                                   0.0  0.0   
                          Boys                                   0.0  0.0   
                          Costumes & Accessories                 0.0  3.9   
                          Men                                    0.0  0.0   
                          Novelty & More                         0.0  0.0   
                          Sport Specific Clothing                0.0  0.0   
                          Women                                  0.0  0.0   
Electronics               Computers & Accessories                4.0